api/traces/v1/trace_rbac.go

```diff
@@ -0,0 +1,182 @@
+			}
+			traceQL = traceQL + " | select(resource.k8s.namespace.name)"
+			query.Set("q", traceQL)
+			r.URL.RawQuery = query.Encode()
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func responseRBACModifier(log log.Logger) func(response *http.Response) error {
+	return func(response *http.Response) error {
+		allowedNamespaces := map[string]bool{}
+		namespaces := apilogsv1.AllowedNamespaces(response.Request.Context())
+		for _, ns := range namespaces {
+			allowedNamespaces[ns] = true
+		}
+		level.Debug(log).Log("AllowedNamespaces", allowedNamespaces)
+
+		if strings.Contains(response.Request.URL.Path, "/api/traces/") || strings.Contains(response.Request.URL.Path, "/api/search") {
+			if response.StatusCode == http.StatusOK {
+				// Uncompressed reader
+				var reader io.ReadCloser
+				var err error
+
+				// Read what Jaeger UI sent back (which might be compressed)
+				switch response.Header.Get("Content-Encoding") {
+				case "gzip":
+					reader, err = gzip.NewReader(response.Body)
+					if err != nil {
+						return err
+					}
+					defer reader.Close()
+				case "deflate":
+					reader = flate.NewReader(response.Body)
+					defer reader.Close()
+				default:
+					reader = response.Body
+				}
+
+				b, err := io.ReadAll(reader)
+				if err != nil {
+					return err
+				}
+				strResponse := string(b)
+				buf := bytes.NewBufferString(strResponse)
+				response.Header["Content-Length"] = []string{fmt.Sprint(buf.Len())}
+				response.Body = io.NopCloser(buf)
+
+				if strings.Contains(response.Request.URL.Path, "/api/traces/") {
+					trace := &tempopb.Trace{}
+					err = tempopb.UnmarshalFromJSONV1(b, trace)
+					if err != nil {
+						return err
+					}
+					trace = traceRBAC(allowedNamespaces, trace)
+
+					traceResponseBody, err := tempopb.MarshalToJSONV1(trace)
+					if err != nil {
+						return err
+					}
+					response.Body = io.NopCloser(bytes.NewBuffer(traceResponseBody))
+				} else {
+					searchResponse := &tempopb.SearchResponse{}
+					err = jsonpb.UnmarshalString(string(b), searchResponse)
+					if err != nil {
+						return err
+					}
+					searchResponse = searchResponseRBAC(allowedNamespaces, searchResponse)
+
+					marshaller := jsonpb.Marshaler{}
+					responseBuffer := &bytes.Buffer{}
+					err = marshaller.Marshal(responseBuffer, searchResponse)
+					if err != nil {
+						return err
+					}
+					response.Body = io.NopCloser(responseBuffer)
+				}
+				// We could re-encode in gzip/deflate, but there is no need, so send it raw
+				response.Header["Content-Encoding"] = []string{}
+			}
+
+			return nil
+		}
+
+		return nil
+	}
+}
+
+func traceRBAC(allowedNamespaces map[string]bool, trace *tempopb.Trace) *tempopb.Trace {
+	for _, rs := range trace.ResourceSpans {
+		notAllowedNamespace := ""
+		if rs.Resource != nil && rs.Resource.Attributes != nil {
+			for _, resAttr := range rs.Resource.Attributes {
+				if resAttr.Key == "k8s.namespace.name" && !allowedNamespaces[resAttr.Value.GetStringValue()] {
```

Comment:
Traces without the `k8s.namespace.name` attribute should also be filtered, they could leak sensitive data.